cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java

```diff
@@ -0,0 +1,370 @@
+ * <pre>
+ *   gcloud config unset auth/disable_credentials
+ *   gcloud config unset api_endpoint_overrides/spanner
+ * </pre>
+ * <p>
+ * Note: The Cloud Spanner Emulator is intended for local development and testing purposes only. It
+ * does not persist data across sessions and is not suitable for production use.
+ * <p>
+ * NOTE: This test class does not currently extend a base test class {@link MetadataStorageTest} because the
+ * {@link SpannerMetadataStorage} implementation is not yet feature-complete. Once all
+ * required methods are implemented, this class will be updated to ensure full test coverage.
+ */
+public class SpannerMetadataStorageTest {
+
+  private static final String EMULATOR_HOST = "localhost:9010";
+  private static final String PROJECT_ID = "test-project";
+  private static final String INSTANCE_ID = "test-instance";
+  private static final String DATABASE_ID = "test-database";
+
+  private static SpannerMetadataStorage spannermetadatastorage;
+  private static DatabaseAdminClient dbAdminClient;
+  private static DatabaseClient dbClient;
+  private static InstanceAdminClient adminClient;
+  private static Spanner spanner;
+
+  // Metadata table names
+  private static final String METADATA_TABLE = "metadata";
+  private static final String METADATA_PROPS_TABLE = "metadata_props";
+
+  // Indicates whether the Spanner emulator is active, guiding cleanup decisions.
+  private static boolean isEmulatorRunning;
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    // Spanner Emulator host not set. Skipping test.
+    Assume.assumeNotNull(System.getenv("SPANNER_EMULATOR_HOST"));
+
+    // Initialize Spanner service.
+    SpannerOptions options = SpannerOptions.newBuilder()
+      .setEmulatorHost(EMULATOR_HOST)
+      .setProjectId(PROJECT_ID)
+      .build();
+    // If emulator is not running, but the env variable SPANNER_EMULATOR_HOST is still set,
+    // Exception is thrown here. Thus, isEmulatorRunning variable is not set to true.
+    spanner = options.getService();
+    dbAdminClient = spanner.getDatabaseAdminClient();
+    adminClient = spanner.getInstanceAdminClient();
+
+    // Cleanup of Spanner resources is necessary only if the emulator is operational.
+    // Without this check, the test may hang indefinitely due to the asynchronous nature of dropDatabase().
+    isEmulatorRunning = true;
+
+    // Create instance if it doesn't exist.
+    try {
+      adminClient.createInstance(InstanceInfo.newBuilder(InstanceId.of(PROJECT_ID, INSTANCE_ID))
+                                   .setInstanceConfigId(InstanceConfigId.of(PROJECT_ID,
+                                                                            "regional-us-central1"))
+                                   .setDisplayName("Test Instance")
+                                   .build()).get();
+    } catch (ExecutionException e) {
+      // Instance already exists, proceed.
```

Comment:
Why is this being assumed?

Reply:
Can we check the cause of the ExecutionException to confirm that instance already exists?